### PR TITLE
Google Pay support, redux

### PR DESCRIPTION
**Pull request by bdewater**
This PR supersedes https://github.com/spreedly/r2d2/pull/10 and uses a composition instead of inheritance to share functions common to both Android and Google Pay. It is definitely easier to follow what's going on 😄 

Also implemented a suggested factory to build the droids... err tokens... based on input.

Once we get this merged I'll spend some time writing additional documentation regarding the verification keys for production use.

### CODE REVIEW

File: lib/r2d2/android_pay_token.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+      shared_secret = generate_shared_secret(private_key, ephemeral_public_key)
+
+      # derive the symmetric_encryption_key and mac_key
+      hkdf_keys = derive_hkdf_keys(ephemeral_public_key, shared_secret, 'Android')
+
+      # verify the tag is a valid value
+      verify_mac(hkdf_keys[:mac_key], encrypted_message, tag)
+
+      JSON.parse(decrypt_message(encrypted_message, hkdf_keys[:symmetric_encryption_key]))
```
Comment (bdewater) [inline lib/r2d2/android_pay_token.rb]: I kept the JSON parse here. In addition to my arguments outlined in https://github.com/spreedly/r2d2/pull/10#discussion_r188063881 - now with the factory method:

```ruby
R2D2.build_token(android_or_google_pay_attrs, recipient_id_etc).decrypt(private_key_pem) 
=> {"dpan"=>"4895370012003478", "expirationMonth"=>12} # etc, always a hash
```

Comment (dtykocki) [reply lib/r2d2/android_pay_token.rb]: Agreed, using the built 'json' library is fine.

**Comment by deedeelavinder** [comment]
👏 Looks good!  👍

**Comment by bdewater** [comment]
party hard

I don't have merge permissions here though 😅  so if you would be so kind to merge and release? ❤️

**Comment by deedeelavinder** [comment]
Yes!!

**Comment by deedeelavinder** [comment]
All commits from this PR have been merged to master. It was decided that `hkdf` be re-added as a runtime-dependency to prevent unintentional failures ( https://github.com/spreedly/r2d2/commit/313ec5103eebe8134439b9a9d97e4548107620b7 )  Fortunately, OpenSSL will still be utilized when available.

**Pull request closed without merging**